keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -182,49 +182,36 @@ updates.
-
-Several core Kubernetes components must be modified (see the
-[implementation](#implementation) section for details):
-- kube-apiserver (new API)
-- kube-controller-manager (new controller)
-- kube-scheduler (new builtin scheduler plugin)
-- kubelet (new third-party kubelet plugin type)
-
-Resources as described by this KEP are managed by third-party resource drivers that communicate with central
-Kubernetes components, in particular the kube-scheduler, by updating
-objects stored in the kube-apiserver. kube-scheduler only needs to be modified
-once to support dynamic resource allocation. Then multiple drivers from
-different vendors can be installed at the same time without making further
-changes to the scheduler.
-
-The scope of this KEP is smaller compared to two previous proposals (linked to
-below under non-goals) and therefore it is easier to implement:
-- kube-scheduler doesn't need to know about different kinds of resources
-  and therefore we don't need to define those in advance and later
-  update those definitions as new hardware appears.
-- There is no need to define standardized parameters for resources.
-
-Communication between the kubelet and the node part of the driver is
-handled through local Unix domain sockets and the kubelet plugin registration
-mechanism, using a new plugin type and a new gRPC interface.
-The container runtime uses the
-[Container Device Interface
-(CDI)](https://github.com/container-orchestrated-devices/container-device-interface)
-to expose the resources.
+Users are increasingly deploying Kubernetes as management solution for new
+workloads (batch processing) and in new environments (edge computing). Such
+workloads no longer need just RAM and CPU, but also access to specialized
+hardware. With upcoming enhancements of data center interconnects, accelerators
+no longer need to be physically plugged into the nodes where they are
+used. Instead, accelerators can be added to nodes dynamically as needed.
+
+This KEP introduces a new API for describing which of these new resource kinds
+a pod needs. The API supports:
+
+- Network-attached resources. The existing [device manager API](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/resource-management/device-plugin.md)
+  is limited to hardware on a node.
+- Sharing of a resource allocation between multiple containers or pods.
+  The device manager API currently cannot share resources at all. It
+  could be extended to share resources between containers in a single pod,
+  but supporting sharing between pods would need a completely new
+  API similar to the one in this KEP.
+- Using a resource that is expensive to initialize multiple times
+  in different pods. This is not possible at the moment.
+- Custom parameters that describe resource requirements and initialization.
+  With the current Kubernetes API, only pod annotations can be used
+  as replacement for such parameters, which is not user-friendly.
```

Comment:
How pod annotations can be used? As far as I remember pod attributes(including annotations) are not passed down to the device plugin on Allocate call and can't be used for passing resource requirements.

Reply:
They can be used by custom CSI drivers. It's very awkward and not at all a good design, but it is doable.

I can add "which is not user-friendly, hard to access from add-ons like CSI drivers and impossible to use for device plugins."

But "impossible" is relative - it's not *currently* supported, but the device plugin API could be modified to support it, so it's not completely impossible.

Reply:
I thought it was about Device Plugins API. I'm not sure we need to compare proposed solution with CSI. Device Plugins API doesn't support in its current state it and it should be explicitly said here.

Reply:
However, we can mention that there were attempts to bring this functionality to DP API if needed.

Reply:
I was thinking of "Kubernetes *Pod* API", not of any particular extension mechanism. All of those mechanisms have the same API problem.

If we compare, then we need to compare against everything that is currently being done, including CSI. The use case from @teozkr is based on the volume API and he explicitly mentioned using annotations as being part of the custom API.

Reply:
New proposed wording:

- Custom parameters that describe resource requirements and initialization.
   With the current Pod API, annotations have to be used to capture such
   parameters and then hacks are needed to access them from a CSI driver
   or device plugin.

Reply:
sounds good to me. thank you.